### PR TITLE
Fixes #741 Diagnostics for empty values in Persistence Annotations

### LINKED ISSUE #741: Empty Value Arrays in Persistence Annotations (@TableGenerator, @SequenceGenerators, @SecondaryTables)
#### Description:
Jakarta Persistence requires certain annotations to declare non‑empty arrays or values. Declaring them with empty arrays or missing values is invalid and leads to misconfigured persistence metadata. 
This diagnostic should detect and report cases where: 

@TableGenerator has no name value specified. 
-A unique generator name that can be referenced by one or more classes to be the generator for primary key values.
@TableGenerators
-The table generator mappings

@SequenceGenerator
-A unique generator name that can be referenced by one or more classes to be the generator for primary key values.
@SequenceGenerators has no sequence generators specified. 
-The sequence generator mappings

@SecondaryTable
-The name of the table.
@SecondaryTables has no secondary tables specified.
-The secondary tables that are used to map the entity class. 

#### Specification:
https://jakarta.ee/specifications/persistence/3.0/jakarta-persistence-spec-3.0#tablegenerators-annotation:~:text=TableGenerator%20Annotation%20Elements
https://jakarta.ee/specifications/persistence/3.0/jakarta-persistence-spec-3.0#tablegenerators-annotation:~:text=The%20table%20generator%20mappings
https://jakarta.ee/specifications/persistence/3.0/jakarta-persistence-spec-3.0#tablegenerators-annotation:~:text=SequenceGenerator%20Annotation%20Elements
https://jakarta.ee/specifications/persistence/3.0/jakarta-persistence-spec-3.0#tablegenerators-annotation:~:text=The%20sequence%20generator%20mappings
https://jakarta.ee/specifications/persistence/3.0/jakarta-persistence-spec-3.0#tablegenerators-annotation:~:text=(Required)-,The%20name%20of%20the%20table.,-String
https://jakarta.ee/specifications/persistence/3.0/jakarta-persistence-spec-3.0#tablegenerators-annotation:~:text=The%20secondary%20tables%20that%20are%20used%20to%20map%20the%20entity%20class

_Select all that apply_
- [x] diagnostic
- [ ] quick-fix
- [ ] snippet
- [ ] other, please specify: 

#### Invalid Usage (should trigger diagnostic): 

```
@Entity 
@TableGenerator() // No generators specified 
public class Customer { 
    @Id 
    private Long id; 
} 

@Entity 
@SequenceGenerators({}) //  Empty array 
public class Order { 
    @Id 
    private Long id; 
} 

@Entity 
@SecondaryTables({}) //  No secondary tables specified 
public class Product { 
    @Id 
    private Long id; 
} 
```

#### Valid Usage: 

```
@Entity @TableGenerator(name = "custGen", table = "id_gen", pkColumnName = "gen_name", valueColumnName = "gen_val") public class Customer { @Id private Long id; } 
@Entity 
@SequenceGenerators({ 
    @SequenceGenerator(name = "orderSeq", sequenceName = "ORDER_SEQ") 
}) 
public class Order { 
    @Id 
    private Long id; 
} 
@Entity 
@SequenceGenerators({ 
    @SequenceGenerator(name = "orderSeq", sequenceName = "ORDER_SEQ") 
}) 
public class Order { 
    @Id 
    private Long id; 
} 
@Entity 
@SecondaryTables({ 
    @SecondaryTable(name = "ProductDetails") 
}) 
public class Product { 
    @Id 
    private Long id; 
} 
```

### PR DESCRIPTION
**Pull request by archana-1924**
Fixes #741

### CODE REVIEW

**Comment by archana-1924** [comment]
https://github.com/user-attachments/assets/4f3b62a5-f78b-49de-b380-457a2a5965c4

File: jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/internal/persistence/PersistenceGeneratorDiagnosticsParticipant.java [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+     * @param diagnostics the mutable list to which a diagnostic is appended on failure
+     * @param messageKey message bundle key for the diagnostic message
+     * @param errorCode error code to attach to the diagnostic
+     * @throws JavaModelException if the annotation's member value pairs cannot be read
+     */
+    private void validateNameAttribute(IAnnotation annotation, JavaDiagnosticsContext context,
+                                       String uri, List<Diagnostic> diagnostics,
+                                       String messageKey, ErrorCode errorCode) throws JavaModelException {
+        String name = DiagnosticUtils.getAnnotationMemberValue(annotation, Constants.NAME, String.class);
+        if (name == null || name.trim().isEmpty()) {
```
Comment (Rejoice472) [inline jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/internal/persistence/PersistenceGeneratorDiagnosticsParticipant.java]: better use this approach 
if (name == null || name.isBlank()) {

**Review by Rejoice472 — APPROVED** [review]

File: jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/internal/persistence/PersistenceGeneratorDiagnosticsParticipant.java [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+ *                 <ul>
+ *                 <li>@TableGenerator must specify a non-empty 'name' attribute</li>
+ *                 <li>@TableGenerators must specify at least one @TableGenerator</li>
+ *                 <li>@SequenceGenerator must specify a non-empty 'name' attribute</li>
+ *                 <li>@SequenceGenerators must specify at least one @SequenceGenerator</li>
+ *                 <li>@SecondaryTable must specify a non-empty 'name' attribute</li>
+ *                 <li>@SecondaryTables must specify at least one @SecondaryTable</li>
+ *                 </ul>
+ */
+public class PersistenceGeneratorDiagnosticsParticipant implements IJavaDiagnosticsParticipant {
```
Comment (Joseph-Bineesh) [inline jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/internal/persistence/PersistenceGeneratorDiagnosticsParticipant.java]: Any reason to go for new participant?

File: jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/internal/persistence/PersistenceGeneratorDiagnosticsParticipant.java [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+     * @param type the enclosing type, used for import resolution
+     * @param context the diagnostics context
+     * @param uri the document URI, used when creating diagnostics
+     * @param diagnostics the mutable list to which any new diagnostics are appended
+     */
+    private void validateAnnotation(IAnnotation annotation, IType type, JavaDiagnosticsContext context,
+                                    String uri, List<Diagnostic> diagnostics) {
+        try {
+            String name = annotation.getElementName();
+            if (DiagnosticUtils.isMatchedJavaElement(type, name, Constants.TABLEGENERATOR)) {
```
Comment (Joseph-Bineesh) [inline jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/internal/persistence/PersistenceGeneratorDiagnosticsParticipant.java]: Can we use switch ? If we use getMatchedJavaElementName with a list of FQN.

File: jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/internal/persistence/PersistenceGeneratorDiagnosticsParticipant.java [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+     * @param uri the document URI, used when creating the diagnostic
+     * @param diagnostics the mutable list to which a diagnostic is appended on failure
+     * @param messageKey message bundle key for the diagnostic message
+     * @param errorCode error code to attach to the diagnostic
+     * @throws JavaModelException if the annotation's member value pairs cannot be read
+     */
+    private void validateNameAttribute(IAnnotation annotation, JavaDiagnosticsContext context,
+                                       String uri, List<Diagnostic> diagnostics,
+                                       String messageKey, ErrorCode errorCode) throws JavaModelException {
+        String name = DiagnosticUtils.getAnnotationMemberValue(annotation, Constants.NAME, String.class);
```
Comment (Joseph-Bineesh) [inline jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/internal/persistence/PersistenceGeneratorDiagnosticsParticipant.java]: Use a better variable name

File: jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/internal/persistence/PersistenceGeneratorDiagnosticsParticipant.java [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+     * @param emptyMappingCode error code for the empty-array diagnostic
+     * @param emptyNameMappingMsgKey message key for an empty {@code name} on a nested annotation
+     * @param emptyNameMappingCode error code for an empty {@code name} on a nested annotation
+     * @throws JavaModelException if the annotation's member value pairs cannot be read
+     */
+    private void validateNonEmptyMappingArray(IAnnotation annotation, JavaDiagnosticsContext context,
+                                              String uri, List<Diagnostic> diagnostics,
+                                              String emptyMappingMsgKey, ErrorCode emptyMappingCode,
+                                              String emptyNameMappingMsgKey, ErrorCode emptyNameMappingCode) throws JavaModelException {
+        Object value = DiagnosticUtils.getAnnotationMemberValue(annotation, "value", Object.class);
```
Comment (Joseph-Bineesh) [inline jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/internal/persistence/PersistenceGeneratorDiagnosticsParticipant.java]: Use better variable name

**Review by Joseph-Bineesh — CHANGES_REQUESTED** [review]

File: jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/internal/persistence/PersistenceGeneratorDiagnosticsParticipant.java (hunk 2 above)
Comment (archana-1924) [reply jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/internal/persistence/PersistenceGeneratorDiagnosticsParticipant.java]: Yes, as ManagedBeanDiagnosticsParticipant is currently crowded and handles produces and inject scenarios. This handles the new 6 diagnostics neatly.

Comment (Joseph-Bineesh) [reply jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/internal/persistence/PersistenceGeneratorDiagnosticsParticipant.java]: Anyway it's not a CDI module. Not expecting this diagnostic in ManagedBean. Did you check the persistence related participants?

File: jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/internal/persistence/PersistenceGeneratorDiagnosticsParticipant.java [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+        IJDTUtils utils = JDTUtilsLSImpl.getInstance();
+        ICompilationUnit unit = utils.resolveCompilationUnit(uri);
+        List<Diagnostic> diagnostics = new ArrayList<>();
+
+        if (unit == null) {
+            return diagnostics;
+        }
+
+        IType[] allTypes = unit.getAllTypes();
+        for (IType type : allTypes) {
```
Comment (Joseph-Bineesh) [inline jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/internal/persistence/PersistenceGeneratorDiagnosticsParticipant.java]: Should we gate check for @Entity annotation? I think we already have this kind of participants. And please note that of participants more than 1000 lines of code not recommend to reuse.

**Review by Joseph-Bineesh — CHANGES_REQUESTED** [review]

File: jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/internal/persistence/PersistenceGeneratorDiagnosticsParticipant.java (hunk 6 above)
Comment (archana-1924) [reply jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/internal/persistence/PersistenceGeneratorDiagnosticsParticipant.java]: As per spec, these can work as standalone annotations. But it only makes sense to use it in @Entity classes

**Review by Joseph-Bineesh — APPROVED** [review]
LGTM

File: jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/internal/persistence/PersistenceEntityDiagnosticsParticipant.java [hunk 7]
```diff
@@ -544,7 +550,130 @@ private void validateFieldOrPropertyType(IMember member, IType type, List<Diagno
...
+     * @param emptyMappingCode error code for the empty-array diagnostic
+     * @param emptyNameMappingMsgKey message key for an empty {@code name} on a nested annotation
+     * @param emptyNameMappingCode error code for an empty {@code name} on a nested annotation
+     * @throws JavaModelException if the annotation's member value pairs cannot be read
+     */
+    private void validateNonEmptyMappingArray(IAnnotation annotation, JavaDiagnosticsContext context,
+                                              String uri, List<Diagnostic> diagnostics,
+                                              String emptyMappingMsgKey, ErrorCode emptyMappingCode,
+                                              String emptyNameMappingMsgKey, ErrorCode emptyNameMappingCode) throws JavaModelException {
+        Object mappingArrayValue = DiagnosticUtils.getAnnotationMemberValue(annotation, Constants.VALUE, Object.class);
```
Comment (venmanyarun) [inline jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/internal/persistence/PersistenceEntityDiagnosticsParticipant.java:733]: Please add a unit test or a runtime log-assert confirming @TableGenerators({}) reaches the empty-array branch in the JDT model.

File: jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/internal/persistence/PersistenceEntityDiagnosticsParticipant.java [hunk 8]
```diff
@@ -76,12 +78,12 @@ public List<Diagnostic> collectDiagnostics(JavaDiagnosticsContext context, IProg
...
-        IType[] alltypes;
-        IAnnotation[] allAnnotations;
-
-        alltypes = unit.getAllTypes();
+        IType[] alltypes = unit.getAllTypes();
         for (IType type : alltypes) {
-            allAnnotations = type.getAnnotations();
+            IAnnotation[] allAnnotations = type.getAnnotations();
+
+            // Validate @TableGenerator/s, @SequenceGenerator/s, @SecondaryTable/s at type level
```
Comment (venmanyarun) [inline jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/internal/persistence/PersistenceEntityDiagnosticsParticipant.java]: According to the Jakarta Persistence spec, @TableGenerator and @SequenceGenerator can legally appear on non-entity classes (they are valid globally anywhere the JPA provider picks them up). But @SecondaryTable is strictly an entity-only annotation.

keep the type-level validation consistent with the field/method gates — move it inside the EntityAnnotation != null block,

File: jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/internal/persistence/Constants.java [hunk 9]
```diff
@@ -31,6 +31,19 @@ public class Constants {
...
     public static final String MAPKEYTEMPORAL = "jakarta.persistence.MapKeyTemporal";
 
+    public static final String TABLE_GENERATOR = "jakarta.persistence.TableGenerator";
+    public static final String TABLE_GENERATORS = "jakarta.persistence.TableGenerators";
+    public static final String SEQUENCE_GENERATOR = "jakarta.persistence.SequenceGenerator";
+    public static final String SEQUENCE_GENERATORS = "jakarta.persistence.SequenceGenerators";
+    public static final String SECONDARY_TABLE = "jakarta.persistence.SecondaryTable";
+    public static final String SECONDARY_TABLES = "jakarta.persistence.SecondaryTables";
+
+    public static final String[] GENERATOR_ANNOTATIONS = {
```
Comment (venmanyarun) [inline jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/internal/persistence/Constants.java]: cleaner name would be GENERATOR_AND_TABLE_ANNOTATIONS

File: jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/internal/persistence/PersistenceEntityDiagnosticsParticipant.java [hunk 10]
```diff
@@ -544,7 +550,130 @@ private void validateFieldOrPropertyType(IMember member, IType type, List<Diagno
...
+        try {
+            String matchedAnnotation = DiagnosticUtils.getMatchedJavaElementName(type, annotation.getElementName(),
+                                                                                 Constants.GENERATOR_ANNOTATIONS);
+            if (matchedAnnotation == null) {
+                return;
+            }
+            switch (matchedAnnotation) {
+                case Constants.TABLE_GENERATOR:
+                    validateGeneratorNameAttribute(annotation, context, uri, diagnostics,
+                                                   "TableGeneratorInvalidEmptyName", ErrorCode.TableGeneratorInvalidEmptyName);
```
Comment (venmanyarun) [inline jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/internal/persistence/PersistenceEntityDiagnosticsParticipant.java]: cant we use errorCode.name() instead of hardcoding "TableGeneratorInvalidEmptyName"

consider all the other places where it applies

File: jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/internal/persistence/PersistenceEntityDiagnosticsParticipant.java (hunk 7 above)
Comment (venmanyarun) [reply jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/internal/persistence/PersistenceEntityDiagnosticsParticipant.java:733]: add for LTI as well

File: jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/internal/persistence/PersistenceEntityDiagnosticsParticipant.java (hunk 8 above)
Comment (venmanyarun) [reply jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/internal/persistence/PersistenceEntityDiagnosticsParticipant.java]: add for LTI as well

File: jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/internal/persistence/Constants.java (hunk 9 above)
Comment (venmanyarun) [reply jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/internal/persistence/Constants.java]: correct in LTI as well

**Review by venmanyarun — CHANGES_REQUESTED** [review]

**Review by venmanyarun — APPROVED** [review]
